File: sdk/nodejs/lib/zbrt-connection.js
/**
 * One ZBRT v1 session over a TCP socket (PROTOCOL.md §3.4, mirroring
 * `zeroboot_connection.rs` / the Java ZbrtConnection): fresh 128-bit request
 * id per request, Output frames strictly before exactly one terminal frame,
 * idempotent Cancel with empty-payload CancelAck, Error frames raise
 * RemoteError. INTERNAL.
 */
import net from 'node:net';
import { randomBytes } from 'node:crypto';
import { DecodeError, RemoteError, TransportError } from './errors.js';
import {
  KIND_CANCEL,
  KIND_CANCEL_ACK,
  KIND_ERROR,
  KIND_EXECUTE,
  KIND_EXIT,
  KIND_FS,
  KIND_FS_RESULT,
  KIND_HELLO,
  KIND_HELLO_ACK,
  KIND_HEALTH,
  KIND_HEALTH_ACK,
  KIND_OUTPUT,
  ZbrtFrame,
  frameReader,
} from './zbrt-frame.js';
import * as codec from './zbrt-codec.js';

export const V1_CAPABILITIES = Object.freeze([
  'execute',
  'stream',
  'deadline',
  'health',
  'cancel',
  'filesystem',
]);

function connectSocket(host, port, timeoutMs) {
  return new Promise((resolve, reject) => {
    const socket = net.createConnection({ host, port });
    const timer = setTimeout(() => {
      socket.destroy();
      reject(new TransportError('guest connection timed out'));
    }, timeoutMs);
    socket.once('connect', () => {
      clearTimeout(timer);
      socket.setNoDelay(true);
      socket.setTimeout(0);
      resolve(socket);
    });
    socket.once('error', (error) => {
      clearTimeout(timer);
      reject(new TransportError(`guest connect failed: ${error.message}`));
    });
    socket.once('timeout', () => {
      clearTimeout(timer);
      socket.destroy();
      reject(new TransportError('guest connection timed out'));
    });
  });
}

/** Reader for one active Execute turn on a ZBRT connection. */
class ZbrtStreamSession {
  constructor(connection, requestId) {
    this.connection = connection;
    this.requestId = requestId;
    this.stopped = false;
    this.terminal = false;
  }

  /** Next event: {stream, data, code} — code != null marks the Exit. */
  async nextEvent() {
    if (this.terminal) return null;
    const frame = await this.connection.#reader.next();
    await this.connection.#requireId(frame, this.requestId);
    if (frame.kind === KIND_OUTPUT) {
      const output = codec.decodeOutput(frame.payload);
      return { stream: output.stream, data: output.data, code: null };
    }
    if (frame.kind === KIND_EXIT) {
      const exit = codec.decodeExit(frame.payload);
      this.terminal = true;
      return { stream: -1, data: Buffer.alloc(0), code: exit.code };
    }
    if (frame.kind === KIND_ERROR) {
      throw this.connection.#errorFrame(frame);
    }
    throw this.connection.#unexpected(frame, 'Output/Exit');
  }

  /** Idempotent stop: Cancel targeting this request, expecting CancelAck. */
  async stop() {
    if (this.terminal || this.stopped) return;
    this.stopped = true;
    const ack = await this.connection.#cancelRoundTrip(null, this.requestId);
    if (ack.payload.length !== 0) {
      throw new DecodeError('CancelAck payload must be empty');
    }
  }
}

export class ZbrtConnection {
  #socket = null;
  #reader = null;
  #ready;

  constructor(host, port, timeoutMs) {
    this.timeoutMs = timeoutMs;
    this.#ready = (async () => {
      this.#socket = await connectSocket(host, port, timeoutMs);
      this.#reader = frameReader(this.#socket);
    })();
  }

  /** Resolves once the TCP connection is up. */
  async ready() {
    await this.#ready;
    return this;
  }

  newRequestId() {
    return randomBytes(16);
  }

  async #writeFrame(frame) {
    if (!this.#socket) throw new TransportError('connection closed');
    await new Promise((resolve, reject) => {
      this.#socket.write(frame.encode(), (error) => (error ? reject(error) : resolve()));
    });
  }

  async #requireId(frame, requestId) {
    if (!frame.requestId.equals(requestId)) {
      throw new DecodeError('request_id mismatch');
    }
  }

  #errorFrame(frame) {
    const error = codec.decodeError(frame.payload);
    return new RemoteError(`${error.code}: ${error.message}`);
  }

  #unexpected(frame, expected) {
    return new DecodeError(`unexpected frame kind ${frame.kind} while awaiting ${expected}`);
  }

  async #cancelRoundTrip(reason, target16) {
    const id = this.newRequestId();
    await this.#writeFrame(new ZbrtFrame(KIND_CANCEL, 0, id, codec.encodeCancel(reason, target16)));
    const frame = await this.#reader.next();
    await this.#requireId(frame, id);
    if (frame.kind === KIND_CANCEL_ACK) return frame;
    if (frame.kind === KIND_ERROR) throw this.#errorFrame(frame);
    throw this.#unexpected(frame, 'CancelAck');
  }

  /** Optional Hello handshake; the guest is auto-ready without it. */
  async hello(clientName) {
    await this.#ready;
    const id = this.newRequestId();
    await this.#writeFrame(
      new ZbrtFrame(KIND_HELLO, 0, id, codec.encodeHello(clientName, V1_CAPABILITIES)),
    );
    const frame = await this.#reader.next();
    await this.#requireId(frame, id);
    if (frame.kind === KIND_HELLO_ACK) {
      return codec.decodeHelloAck(frame.payload);
    }
    throw this.#unexpected(frame, 'HelloAck');
  }

  /**
   * Run one command: collect Output frames (0=stdout, 1=stderr) until exactly
   * one terminal frame — Exit (completed/cancelled) or Error (raise).
   * timeoutMs rides the wire as the guest-side deadline.
   */
  async execute(argv, cwd, stdin, timeoutMs) {
    await this.#ready;
    const id = this.newRequestId();
    await this.#writeFrame(
      new ZbrtFrame(KIND_EXECUTE, 0, id, codec.encodeExecute(argv, cwd, stdin ?? Buffer.alloc(0), timeoutMs)),
    );
    const stdout = [];
    const stderr = [];
    while (true) {
      const frame = await this.#reader.next();
      await this.#requireId(frame, id);
      if (frame.kind === KIND_OUTPUT) {
        const output = codec.decodeOutput(frame.payload);
        (output.stream === 0 ? stdout : stderr).push(output.data);
      } else if (frame.kind === KIND_EXIT) {
        const exit = codec.decodeExit(frame.payload);
        return {
          code: exit.code,
          stdout: Buffer.concat(stdout),
          stderr: Buffer.concat(stderr),
          signal: exit.signal,
          timedOut: false,
        };
      } else if (frame.kind === KIND_ERROR) {
        throw this.#errorFrame(frame);
      } else {
        throw this.#unexpected(frame, 'Output/Exit');
      }
    }
  }

  /** Idempotent cancel; expects the empty-payload CancelAck. */
  async cancel(reason, target16) {
    if (target16 !== null && target16 !== undefined && target16.length !== 16) {
      throw new DecodeError('cancel target must be 16 bytes');
    }
    const frame = await this.#cancelRoundTrip(reason, target16);
    if (frame.payload.length !== 0) {
      throw new DecodeError('CancelAck payload must be empty');
    }
  }

  /** Health round-trip; the guest reports healthy=true, message="ready". */
  async health() {
    await this.#ready;
    const frame = await this.#roundTrip(
      KIND_HEALTH,
      codec.encodeHealth(true, null),
      KIND_HEALTH_ACK,
      'HealthAck',
    );
    return codec.decodeHealth(frame.payload);
  }

  /** Route one filesystem RPC (opcodes 1=ls 2=find 3=grep 4=read 5=write). */
  async fs(op, path, jsonArgs) {
    await this.#ready;
    const frame = await this.#roundTrip(
      KIND_FS,
      codec.encodeFs(op, path, jsonArgs),
      KIND_FS_RESULT,
      'FsResult',
    );
    return frame.payload;
  }

  /**
   * Start an interactive turn: send Execute and return a session that yields
   * Output/Exit frames. Only one turn may be active per connection.
   */
  async openStreamSession(argv, cwd, stdin, timeoutMs) {
    await this.#ready;
    const id = this.newRequestId();
    await this.#writeFrame(
      new ZbrtFrame(KIND_EXECUTE, 0, id, codec.encodeExecute(argv, cwd, stdin ?? Buffer.alloc(0), timeoutMs)),
    );
    return new ZbrtStreamSession(this, id);
  }

  close() {
    this.#socket?.end();
    this.#socket?.destroy();
  }
}

File: sdk/java/src/main/java/io/rfb/sdk/internal/ZbrtConnection.java
package io.rfb.sdk.internal;

import io.rfb.sdk.DecodeError;
import io.rfb.sdk.RemoteError;
import io.rfb.sdk.TransportError;

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.security.SecureRandom;
import java.time.Duration;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * One ZBRT v1 session over a TCP socket (PROTOCOL.md §3.4, mirroring
 * {@code zeroboot_connection.rs} semantics): fresh 128-bit request id per
 * request, Output frames strictly before exactly one terminal frame, idempotent
 * Cancel with empty-payload CancelAck, Error frames raise {@link RemoteError}.
 * INTERNAL.
 */
public final class ZbrtConnection implements AutoCloseable {
    /** ZBRT V1 wire capability vocabulary (single source of truth in the Rust crate). */
    public static final List<String> V1_CAPABILITIES =
            Collections.unmodifiableList(Arrays.asList(
                    "execute", "stream", "deadline", "health", "cancel", "filesystem"));

    /** Shared per-process source: SecureRandom.nextBytes is thread-safe. */
    private static final SecureRandom RANDOM = new SecureRandom();

    private final Socket socket;
    private final InputStream in;
    private final java.io.OutputStream out;

    public ZbrtConnection(InetSocketAddress address, Duration timeout) {
        this.socket = GuestNdjson.connect(address, timeout);
        try {
            this.in = new BufferedInputStream(socket.getInputStream());
            this.out = socket.getOutputStream();
        } catch (IOException e) {
            try {
                socket.close();
            } catch (IOException ignored) {
                // best effort
            }
            throw new TransportError("zbrt stream setup failed: " + e.getMessage(), e);
        }
    }

    /** Fresh 128-bit request id, generated per request. */
    public byte[] newRequestId() {
        byte[] id = new byte[16];
        RANDOM.nextBytes(id);
        return id;
    }

    /** Optional Hello handshake; the guest is auto-ready without it. */
    public ZbrtCodec.HelloAck hello(String clientName) {
        byte[] id = newRequestId();
        writeFrame(new ZbrtFrame(ZbrtFrame.KIND_HELLO, 0, id,
                ZbrtCodec.encodeHello(clientName, V1_CAPABILITIES)));
        ZbrtFrame frame = readReply(id);
        if (frame.kind() == ZbrtFrame.KIND_HELLO_ACK) {
            return ZbrtCodec.decodeHelloAck(frame.payload());
        }
        throw unexpected(frame, "HelloAck");
    }

    /**
     * Run one command: collect Output frames (0=stdout, 1=stderr) until exactly
     * one terminal frame — Exit (completed/cancelled) or Error (raise).
     * {@code timeoutMs} is sent on the wire as the guest-side deadline; a read
     * stall at the connection timeout is a transport-level failure
     * ({@link TransportError}) — the client never cancels on its own (the
     * Rust baseline surfaces timeouts the same way).
     */
    public Exec execute(List<String> argv, String cwd, byte[] stdin, long timeoutMs) {
        byte[] id = newRequestId();
        writeFrame(new ZbrtFrame(ZbrtFrame.KIND_EXECUTE, 0, id,
                ZbrtCodec.encodeExecute(argv, cwd, stdin == null ? new byte[0] : stdin, timeoutMs)));
        ByteArrayOutputStream stdoutBuf = new ByteArrayOutputStream();
        ByteArrayOutputStream stderrBuf = new ByteArrayOutputStream();
        while (true) {
            ZbrtFrame frame = ZbrtFrame.decode(in);
            requireId(frame, id);
            if (frame.kind() == ZbrtFrame.KIND_OUTPUT) {
                ZbrtCodec.Output output = ZbrtCodec.decodeOutput(frame.payload());
                if (output.stream() == 0) {
                    stdoutBuf.write(output.data(), 0, output.data().length);
                } else {
                    stderrBuf.write(output.data(), 0, output.data().length);
                }
            } else if (frame.kind() == ZbrtFrame.KIND_EXIT) {
                ZbrtCodec.Exit exit = ZbrtCodec.decodeExit(frame.payload());
                return new Exec(exit.code(), stdoutBuf.toByteArray(), stderrBuf.toByteArray(),
                        exit.signal(), false);
            } else if (frame.kind() == ZbrtFrame.KIND_ERROR) {
                throw errorFrame(frame);
            } else {
                throw unexpected(frame, "Output/Exit");
            }
        }
    }

    /** Idempotent cancel; expects the empty-payload CancelAck. */
    public void cancel(String reason, byte[] target16) {
        if (target16 != null && target16.length != 16) {
            throw new DecodeError("cancel target must be 16 bytes");
        }
        ZbrtFrame frame = cancelRoundTrip(reason, target16);
        if (frame.payload().length != 0) {
            throw new DecodeError("CancelAck payload must be empty");
        }
    }

    /** Health round-trip; the guest reports healthy=true, message="ready". */
    public ZbrtCodec.Health health() {
        ZbrtFrame frame = roundTrip(ZbrtFrame.KIND_HEALTH,
                ZbrtCodec.encodeHealth(true, null), ZbrtFrame.KIND_HEALTH_ACK, "HealthAck");
        return ZbrtCodec.decodeHealth(frame.payload());
    }

    /** Route one filesystem RPC (opcodes 1=ls 2=find 3=grep 4=read 5=write). */
    public byte[] fs(int op, String path, byte[] jsonArgs) {
        return roundTrip(ZbrtFrame.KIND_FS, ZbrtCodec.encodeFs(op, path, jsonArgs),
                ZbrtFrame.KIND_FS_RESULT, "FsResult").payload();
    }

    /**
     * Start an interactive turn: send Execute and return a reader that yields
     * Output/Exit frames. Only one turn may be active per connection.
     */
    public ZbrtStreamSession openStreamSession(List<String> argv, String cwd, byte[] stdin, long timeoutMs) {
        byte[] id = newRequestId();
        writeFrame(new ZbrtFrame(ZbrtFrame.KIND_EXECUTE, 0, id,
                ZbrtCodec.encodeExecute(argv, cwd, stdin == null ? new byte[0] : stdin, timeoutMs)));
        return new ZbrtStreamSession(id);
    }

    /** Reader for one active Execute turn on this connection. */
    public final class ZbrtStreamSession {
        private final byte[] requestId;
        private volatile boolean stopped = false;
        private volatile boolean terminal = false;

        ZbrtStreamSession(byte[] requestId) {
            this.requestId = requestId;
        }

        /** Next event frame: Output or terminal Exit; null only if we already terminated. */
        public Event nextEvent() {
            if (terminal) {
                return null;
            }
            ZbrtFrame frame = ZbrtFrame.decode(in);
            requireId(frame, requestId);
            if (frame.kind() == ZbrtFrame.KIND_OUTPUT) {
                ZbrtCodec.Output output = ZbrtCodec.decodeOutput(frame.payload());
                return new Event(output.stream(), output.data(), null);
            }
            if (frame.kind() == ZbrtFrame.KIND_EXIT) {
                ZbrtCodec.Exit exit = ZbrtCodec.decodeExit(frame.payload());
                terminal = true;
                return new Event(-1, new byte[0], exit.code());
            }
            if (frame.kind() == ZbrtFrame.KIND_ERROR) {
                throw errorFrame(frame);
            }
            throw unexpected(frame, "Output/Exit");
        }

        /** Idempotent stop: Cancel targeting this request, expecting CancelAck. */
        public void stop() {
            if (terminal || stopped) {
                return;
            }
            stopped = true;
            // sendCancel writes the targeted Cancel and consumes the CancelAck.
            // The guest then delivers this turn's single terminal (Exit) which
            // nextEvent() reports to the caller.
            sendCancel(null, requestId);
        }

        public byte[] requestId() {
            return requestId;
        }
    }

    /** One ZbrtStreamSession event: stream 0=stdout 1=stderr, or code != null → Exit. */
    public static final class Event {
        private final int stream;
        private final byte[] data;
        private final Integer code;

        public Event(int stream, byte[] data, Integer code) {
            this.stream = stream;
            this.data = data;
            this.code = code;
        }

        public int stream() { return stream; }
        public byte[] data() { return data; }
        public Integer code() { return code; }

        public boolean isExit() {
            return code != null;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Event)) return false;
            Event other = (Event) o;
            return stream == other.stream
                    && java.util.Arrays.equals(data, other.data)
                    && java.util.Objects.equals(code, other.code);
        }

        @Override
        public int hashCode() {
            return java.util.Objects.hash(stream, java.util.Arrays.hashCode(data), code);
        }

        @Override
        public String toString() {
            return "Event[stream=" + stream + ", data=" + java.util.Arrays.toString(data)
                    + ", code=" + code + "]";
        }
    }

    public static final class Exec {
        private final int code;
        private final byte[] stdout;
        private final byte[] stderr;
        private final Long signal;
        private final boolean timedOut;

        public Exec(int code, byte[] stdout, byte[] stderr, Long signal, boolean timedOut) {
            this.code = code;
            this.stdout = stdout;
            this.stderr = stderr;
            this.signal = signal;
            this.timedOut = timedOut;
        }

        public int code() { return code; }
        public byte[] stdout() { return stdout; }
        public byte[] stderr() { return stderr; }
        public Long signal() { return signal; }
        public boolean timedOut() { return timedOut; }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Exec)) return false;
            Exec other = (Exec) o;
            return code == other.code
                    && timedOut == other.timedOut
                    && java.util.Arrays.equals(stdout, other.stdout)
                    && java.util.Arrays.equals(stderr, other.stderr)
                    && java.util.Objects.equals(signal, other.signal);
        }

        @Override
        public int hashCode() {
            return java.util.Objects.hash(code, java.util.Arrays.hashCode(stdout),
                    java.util.Arrays.hashCode(stderr), signal, timedOut);
        }

        @Override
        public String toString() {
            return "Exec[code=" + code + ", stdout=" + java.util.Arrays.toString(stdout)
                    + ", stderr=" + java.util.Arrays.toString(stderr)
                    + ", signal=" + signal + ", timedOut=" + timedOut + "]";
        }
    }

    // ---- plumbing --------------------------------------------------------

    private void sendCancel(String reason, byte[] target16) {
        cancelRoundTrip(reason, target16);
    }

    /**
     * Send Cancel and wait for the CancelAck. The guest may emit straggler
     * Output/Exit frames for the cancelled turn before the ack; they are
     * skipped instead of failing the decode (mirrors the Python reference).
     */
    private ZbrtFrame cancelRoundTrip(String reason, byte[] target16) {
        byte[] id = newRequestId();
        writeFrame(new ZbrtFrame(ZbrtFrame.KIND_CANCEL, 0, id,
                ZbrtCodec.encodeCancel(reason, target16)));
        while (true) {
            ZbrtFrame frame = readReply(id);
            if (frame.kind() == ZbrtFrame.KIND_CANCEL_ACK) {
                return frame;
            }
            if (frame.kind() == ZbrtFrame.KIND_ERROR) {
                throw errorFrame(frame);
            }
            if (frame.kind() == ZbrtFrame.KIND_OUTPUT || frame.kind() == ZbrtFrame.KIND_EXIT) {
                continue;
            }
            throw unexpected(frame, "CancelAck");
        }
    }

    /**
     * Send one request frame and read the reply with a matching request id.
     * An Error reply raises {@link RemoteError}; anything other than
     * {@code ackKind} raises {@link DecodeError}.
     */
    private ZbrtFrame roundTrip(int requestKind, byte[] payload, int ackKind, String expected) {
        byte[] id = newRequestId();
        writeFrame(new ZbrtFrame(requestKind, 0, id, payload));
        ZbrtFrame frame = readReply(id);
        if (frame.kind() == ackKind) {
            return frame;
        }
        if (frame.kind() == ZbrtFrame.KIND_ERROR) {
            throw errorFrame(frame);
        }
        throw unexpected(frame, expected);
    }

    private ZbrtFrame readReply(byte[] requestId) {
        ZbrtFrame frame = ZbrtFrame.decode(in);
        requireId(frame, requestId);
        return frame;
    }

    private void requireId(ZbrtFrame frame, byte[] requestId) {
        if (!java.util.Arrays.equals(frame.requestId(), requestId)) {
            throw new DecodeError("request_id mismatch");
        }
    }

    private RemoteError errorFrame(ZbrtFrame frame) {
        ZbrtCodec.ZbrtErrorPayload error = ZbrtCodec.decodeError(frame.payload());
        return new RemoteError(error.code(), error.message());
    }

    private DecodeError unexpected(ZbrtFrame frame, String expected) {
        return new DecodeError("unexpected frame kind " + frame.kind() + " while awaiting " + expected);
    }

    private void writeFrame(ZbrtFrame frame) {
        frame.writeTo(out);
    }

    @Override
    public synchronized void close() {
        try {
            socket.close();
        } catch (IOException ignored) {
            // best effort
        }
    }
}
